sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c

```diff
@@ -246,6 +253,16 @@ size_t gnrc_lorawan_build_uplink(gnrc_lorawan_t *mac, iolist_t *payload,
     lw_hdr->addr = mac->dev_addr;
     lw_hdr->fctrl = 0;
 
+    lorawan_hdr_set_adr(lw_hdr, IS_ACTIVE(CONFIG_LORAMAC_DEFAULT_ADR));
+
+    /* Set `ADRACKReq` bit and ADR_ACK_CNT */
+    if ((mac->last_dr != 0) && IS_ACTIVE(CONFIG_LORAMAC_DEFAULT_ADR)) {
```

Comment:
the `adr_ack_cnt` should also be incremented when sending in DR0

Reply:
Are you sure, this is a grey area in standard

Reply:
The standard says this value should be incremented with every uplink. I don't understand where's the gray zone.

And after all, since all these values are power of 2, it's straightforward to get these "modulo" operation using masks. So it wouldn't add an overhead IMO.

Reply:
But from my understanding ( line 441 in standard ), this process is applicable for 

> an end-device whose data rate is optimized by the network to use a data rate higher than its **lowest available data rate**

Reply:
basically you can use something like these conditions (not tested):
```c
/* Assumming ADR_ACK_COUNTER incremented on every uplink */

if (adr_is_enabled && mac->last_dr != 0) {
    bool should_set_adr_bit =  mac->mlme.adr_ack_cnt > ADR_ACK_LIMIT;
    bool should_decrease_dr = (mac->mlme.adr_ack_cnt - ADR_ACK_LIMIT) % ADR_ACK_DELAY == 0
/* ... */
}
```

Reply:
since ADR_ACK_LIMIT and ADR_ACK_DELAY are most of the times power of 2, the compiler should be able to write efficient code

Reply:
Thanks for the suggestion but my question remains. Should we increment the counter even when DR == 0?

Reply:
Thanks for the suggestion :)

Reply:
no, this is what I described in above in the code. The ADR ACK Counter keeps incrementing (as the standard specifies). The only interesting events are "should_set_adr_bit" and "should_decrease_dr". The ADR Ack Counter can still keep adding up

Reply:
just to summarize: if ADR is on, the ADR counter increments in the same way the uplink counter increments. This means, for every uplink.
It's reset each time there's a downlink

Reply:
But why should we increment the ADR ACK Counter if " DR == 0 ". The standard assumed that the device is not in its  **lowest available data rate**. Does not however state "clearly"  what to do after it reaches the lowest DR. IMHO if the device reached the lowest available DR incrementing ADR ACK Counter does not make sense.

Reply:
The standard says:
```
The ADRACKReq SHALL not be set if the device uses its lowest available data rate
because in that case no action can be taken to improve the link range.
```

This is exactly what we should do IMO. Simply not setting the ADRAckReq bit if `DR == 0`. Otherwise why the hassle of adding a separate logic to treat this special case? IMO all these corner cases are harder to read and error prone.
If you are worried about additional 4 bytes of RAM, [we can already remove quite some in the header files](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/link_layer/lorawan/include/gnrc_lorawan_internal.h#L178)